libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,494 @@
+flash will be the `valid` flag. The object header for ONE will now look like:
+
+```
+0x400                                                                                              0x52C
+--------------------------------------------------------------------------------------------------------
+||||| version|len/flag|   len  |                              hashed_key                               |
+|||||        |        |        |        |        |        |        |        |        |        |        |
+|||||    0x00|00000000|    0x34|    0xed|    0xa1|    0x00|    0x78|    0x88|    0x61|    0x93|    0xbb|
+--------------------------------------------------------------------------------------------------------
+              ^
+```
+
+No changes will happen in flash until key TWO has also been invalidated.
+At which point `garbage_collect()` can erase the region.
+
+## Limitations of TickFS
+
+### Fragmentation
+
+Although TickFS has a `garbage_collect()` function, it makes no effort to
+handle fragmentation.
+
+That means that if you have the following objects in a region
+
+```
+0x000                  0x400
+----------------------------
+|||||     Region 0     |||||
+|||||                  |||||
+|||||ONE|TWO|THREE|FOUR|||||
+----------------------------
+```
+
+Where ONE, TWO and FOUR have been marked as invalid the entire region will still
+be classified as full. It will only not be full when all four objects are marked
+for invalid.
+
+It would be possible to have a reserved region and then free up the region with
+the following steps:
+ 1. Move the region to the reserved region
+ 1. Erase the above region
+ 1. Move the valid objects from the reserved region to the original region,
+    after being defraged in memory
+ 1. Erase the reserved region
+
+The above steps would ensure we don't loose data on a power loss, as long as
+we could ensure we could continue the operation after regaining power.
+
+There are a few problems with this approach though.
+
+In order to maintain "garbage collecting" state to restore after a power loss
+we would need some sore of super block. This would then break the requirement
+on wear leveling as whichever region that block is in will be written and
+erased more often the others.
+
+The other problem is the reserved region where we would move data to. This
+would cost space as an entire region would be dedicated to garbage collecting.
+The garbage collecting region would also have more erase and writes performed
+on it breaking the wear leveling requirement.
+
+### Somewhat high storage overhead
+
+The storage overhead is somewhat high for TickFS. This is mostly due to the
+two 64-bit hashes that are stored with every object.
+
+The 64-bit value is determined by the output of the standard Rust
+`core::hash::Hasher` trait.
+
+For the hash of the key a 64-bit value can be justified by the lack of
+collision avoidance in the implementation. If two keys have the same
```

Comment:
Could you add salt, to avoid collisions? E.g., if you insert and it fails due to a collision, you can retry with a different salt?

Reply:
That's an option, but how do we find the salt when looking up the keys? Do we just have a list of salts to walk through?